crates/core/src/factories/core_fetch.rs

```diff
@@ -0,0 +1,301 @@
+//! Fetch is a Kitsune2 module for fetching ops from peers.
+//!
+//! In particular it tracks which ops need to be fetched from which agents,
+//! sends fetch requests and processes incoming responses to these requests.
+//!
+//! It consists of multiple parts:
+//! - State object that tracks op and agent ids in memory
+//! - Fetch tasks that request tracked ops from agents
+//! - Incoming op task that processes incoming responses to op requests by
+//!     - persisting ops to the data store
+//!     - removing op ids from in-memory data object
+//!
+//! ### State object [CoreFetch]
+//!
+//! - Exposes public method [CoreFetch::add_ops] that takes a list of op ids and an agent id.
+//! - Stores pairs of ([OpId][AgentId]) in a set.
+//! - A hash set` is used to look up elements by key efficiently. Ops may be added redundantly
+//!   to the set with different sources to fetch from, so the set is keyed by op and agent id together.
+//!
+//! ### Fetch tasks
+//!
+//! A channel acts as the queue structure for the fetch tasks. Ops to fetch are sent
+//! one by one through the channel to the receiving tasks running in parallel. The flow
+//! of sending fetch requests is as follows:
+//!
+//! - Await fetch requests for ([OpId], [AgentId]) from an internal channel.
+//! - Check if requested op id/agent id is still on the list of ops to fetch.
+//!     - In case the op has been received in the meantime and no longer needs to be fetched,
+//!       do nothing.
+//!     - Otherwise proceed.
+//! - Check if agent is on a cool-down list of unresponsive agents.
+//! - Dispatch request for op id from agent to transport module.
+//! - If agent is unresponsive, put them on cool-down list.
+//! - Re-send requested ([OpId], [AgentId]) to the internal channel again. It will be removed
+//!   from the list of ops to fetch if it is received in the meantime, and thus prevent a redundant
+//!   fetch request.
+//!
+//! ### Incoming op task
+//!
+//! - Incoming op is written to the data store.
+//! - Once persisted successfully, op is removed from the data object.
+
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Arc,
+    time::Instant,
+};
+
+use kitsune2_api::{
+    builder,
+    config::ModConfig,
+    fetch::{DynFetch, DynFetchFactory, Fetch, FetchFactory},
+    tx::Transport,
+    AgentId, BoxFut, K2Result, OpId,
+};
+use tokio::sync::{
+    mpsc::{channel, Receiver, Sender},
+    Mutex,
+};
+
+const MOD_NAME: &str = "Fetch";
+
+/// Configuration parameters for [CoreFetchFactory].
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct CoreFetchConfig {
+    /// How many parallel op fetch requests can be made at once. Default: 2.  
+    pub parallel_request_count: u8,
+    /// Duration in ms to keep an unresponsive agent on the cool-down list. Default: 10_000.
+    pub cool_down_interval_ms: u64,
```

Comment:
done #259aac531c78dfe62df2d067dd636b28f249a284

Reply:
Hmm, I mean that is better in the case that the agent is gone. It's probably still okay in the case of an interrupted connection. It would be less good in the case of a flaky connection.

I think we should come back to this later and put in a full exponential backoff

Reply:
Ah, that's the difference between cool-down and back-off for you. Perhaps an exponential backoff is better. It would mean that there is repeated successful gossip with the peer, but when fetching ops the connection is flaky, then good again when gossiping. A problem when used maliciously.

Reply:
This will be joined up by the peer behavior module. I agree we should be tracking connectivity across modules and not having each module make different decisions.